### ISSUE TITLE
[prefer-readonly] False positive on fields within mixin classes

### ISSUE THREAD

**Issue by WGroenestein** [typescript-eslint/typescript-eslint#2590]
**Repro**
Using the repo code from https://github.com/WGroenestein/typescript-eslint-prefer-readonly-class-mixin

1. run "npm install"
2. run "npm run build"
3. see that the build compiles without errors
4. run "npm run lint"
5. see that it fails with error
```log
file.ts
  7:3  error  Member '_name' is never reassigned; mark it as `readonly`  @typescript-eslint/prefer-readonly

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

```
6. run "npm run fix-lint"
7. see that it prefixed "_name" in "file.ts" with readonly
8. run "npm run build"
9. see that the build fails with error
```log
file.ts(10,9): error TS2540: Cannot assign to '_name' because it is a read-only property.
```
**Expected Result**

Fields which are assigned outside the constructor (e.g. setters) in mixin classes are not marked as not assigned / readonly

**Actual Result**

They are reported as errors, and when running --fix it breaks the build

**Additional Info**

The debug log can be found in the repo (https://github.com/WGroenestein/typescript-eslint-prefer-readonly-class-mixin/blob/master/debug.log)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.2.0` |
| `@typescript-eslint/parser`        | `4.2.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `12.18.3` |

**Comment by bradzacher** [comment]
Had a quick look and the bug occurs due to the type of the class that TS reports.
It reports that the type of the parent class is a dynamic type: `{new (...args: any[]) => T} & TBase`
This doesn't have a symbol attached, so the tooling just skips it, hence it messes up.

**Comment by sosoba** [comment]
Another example: 
```ts
export class Foo {

  constructor(private bar: string) {
  }

  getBar(){
    return this.bar;
  }

  setBar(){
    this.bar = '';
  }

}
```
Rule says: ` 3:15  warning  Member 'bar: string' is never reassigned; mark it as 'readonly'` which is not true .

**Comment by armano2** [comment]
this seem like issue with types produced by typescript, generally i do not like that this rule is using type checking, as there should no be a need for it, we could achieve all of this by doing static code analysis, 

```
function ClassWithName<TBase extends new (...args: any[]) => {}>(Base: TBase) {
	return class extends Base {

		private _name: string;

		public test(value: string) {
			this._name = value;
		}
	}
}
```

in this case, typescript returns us IntersectionType without symbol

in typeFlags we are receiving `Intersection = 2097152` instead of `Object = 524288`, we can simply fix it by adding handling to this case, but we should def report this to typescript team
